src/test/test-fs-util.c

```diff
@@ -1245,6 +1246,47 @@ TEST(xopenat) {
         fd = safe_close(fd);
 }
 
+TEST(xopenat_lock) {
+        _cleanup_(rm_rf_physical_and_freep) char *t = NULL;
+        _cleanup_close_ int tfd = -EBADF, fd = -EBADF;
+        siginfo_t si;
+
+        assert_se((tfd = mkdtemp_open(NULL, 0, &t)) >= 0);
+
+        /* Test that we can acquire an exclusive lock on a directory in one process, remove the directory,
+         * and close the file descriptor and still properly create the directory and acquire the lock in
+         * another process.  */
+
+        fd = xopenat_lock(tfd, "abc", O_CREAT|O_DIRECTORY|O_CLOEXEC, 0755, LOCK_BSD, LOCK_EX);
+        assert_se(fd >= 0);
+        assert_se(faccessat(tfd, "abc", F_OK, 0) >= 0);
+        assert_se(fd_verify_directory(fd) >= 0);
+        assert_se(xopenat_lock(tfd, "abc", O_DIRECTORY|O_CLOEXEC, 0755, LOCK_BSD, LOCK_EX|LOCK_NB) == -EAGAIN);
+
+        pid_t pid = fork();
+        assert_se(pid >= 0);
+
+        if (pid == 0) {
+                safe_close(fd);
+
+                fd = xopenat_lock(tfd, "abc", O_CREAT|O_DIRECTORY|O_CLOEXEC, 0755, LOCK_BSD, LOCK_EX);
+                assert_se(fd >= 0);
+                assert_se(faccessat(tfd, "abc", F_OK, 0) >= 0);
+                assert_se(fd_verify_directory(fd) >= 0);
+                assert_se(xopenat_lock(tfd, "abc", O_DIRECTORY|O_CLOEXEC, 0755, LOCK_BSD, LOCK_EX|LOCK_NB) == -EAGAIN);
+
+                _exit(0);
+        }
+
+        assert_se(usleep(20 * USEC_PER_MSEC) >= 0);
```

Comment:
That would change the test, we need to give the subprocess time to open the directory and start wait for the lock, so we need to wait a bit. The test will work regardless of which order the logic exexutes.

Reply:
ah ok, I misunderstood it then

Reply:
wouldn't it be nicer if we'd use an eventfd to sync this? i.e. when fork acquired the lock, signal the eventfd and wait for it in the parent instead of this usleep()?

Reply:
It will never acquire the lock though, we need the fork to be blocked in xopenat_lock() in lock_generic() waiting for the lock, then we remove the directory and close the file descriptor to trigger the fstat() logic in xopenat_lock(). I'm not sure how to do that with an eventfd. We'd need to signal it after the open in xopenat_lock() but obviously we can't do that.